source/utilities.cc

```diff
@@ -247,117 +247,291 @@ namespace aspect
 
 
 
-    namespace
+    namespace MapParsing
     {
-      // This is a helper function used in parse_map_to_double_array below.
-      // It takes an input_string that is expected to follow the input format
-      // explained in the documentation of the parse_map_to_double_array function
-      // and parses it into a multimap, only performing rudimentary error checking
-      // for correct formatting.
-      std::multimap<std::string, double>
-      parse_string_to_map (const std::string &input_string,
-                           const std::vector<std::string> &list_of_keys,
-                           const std::string &property_name)
+      namespace
       {
-        std::multimap<std::string, double> parsed_map;
+        // This is a helper function used in parse_map_to_double_array below.
+        // It takes an input_string that is expected to follow the input format
+        // explained in the documentation of the parse_map_to_double_array function
+        // and parses it into a multimap, only performing rudimentary error checking
+        // for correct formatting.
+        std::multimap<std::string, double>
+        parse_string_to_map (const std::string &input_string,
+                             const Options &options)
+        {
+          std::multimap<std::string, double> parsed_map;
 
-        if (Patterns::Map(Patterns::Anything(),
-                          Patterns::List(Patterns::Double(),
-                                         0,
-                                         std::numeric_limits<unsigned int>::max(),
-                                         "|")).match(input_string))
-          {
-            // Split the list by comma delimited components.
-            const std::vector<std::string> field_entries = dealii::Utilities::split_string_list(input_string, ',');
+          if (Patterns::Map(Patterns::Anything(),
```

Comment:
I think it might be worth adding a comment block up here describing the logic and differences between the different cases (Anything, list of allowed, list of required)